parquet/src/main/java/org/apache/iceberg/data/parquet/InternalReader.java

```diff
@@ -0,0 +1,130 @@
+import org.apache.iceberg.parquet.ParquetValueReader;
+import org.apache.iceberg.parquet.ParquetValueReaders;
+import org.apache.iceberg.parquet.ParquetValueReaders.StructReader;
+import org.apache.iceberg.types.Types.StructType;
+import org.apache.parquet.column.ColumnDescriptor;
+import org.apache.parquet.schema.LogicalTypeAnnotation;
+import org.apache.parquet.schema.MessageType;
+import org.apache.parquet.schema.Type;
+
+public class InternalReader extends BaseParquetReaders<StructLike> {
+
+  private static final InternalReader INSTANCE = new InternalReader();
+
+  private InternalReader() {}
+
+  public static ParquetValueReader<StructLike> buildReader(
+      Schema expectedSchema, MessageType fileSchema) {
+    return INSTANCE.createReader(expectedSchema, fileSchema);
+  }
+
+  public static ParquetValueReader<StructLike> buildReader(
+      Schema expectedSchema, MessageType fileSchema, Map<Integer, ?> idToConstant) {
+    return INSTANCE.createReader(expectedSchema, fileSchema, idToConstant);
+  }
+
+  @Override
+  protected ParquetValueReader<StructLike> createStructReader(
+      List<Type> types, List<ParquetValueReader<?>> fieldReaders, StructType structType) {
+    return new StructLikeReader(types, fieldReaders, structType);
+  }
+
+  @Override
+  protected ParquetValueReaders.PrimitiveReader<?> fixedReader(ColumnDescriptor desc) {
+    return new ParquetValueReaders.BytesReader(desc);
+  }
+
+  @Override
+  protected ParquetValueReaders.PrimitiveReader<?> int96Reader(ColumnDescriptor desc) {
+    return new ParquetValueReaders.TimestampInt96Reader(desc);
+  }
+
+  @Override
+  protected Optional<ParquetValueReader<?>> dateReader(ColumnDescriptor desc) {
+    return Optional.of(new ParquetValueReaders.UnboxedReader<>(desc));
+  }
+
+  @Override
+  protected Optional<ParquetValueReader<?>> timeReader(
+      ColumnDescriptor desc, LogicalTypeAnnotation.TimeUnit unit) {
+    if (unit == LogicalTypeAnnotation.TimeUnit.MILLIS) {
+      return Optional.of(new ParquetValueReaders.TimestampMillisReader(desc));
+    }
+
+    return Optional.of(new ParquetValueReaders.UnboxedReader<>(desc));
+  }
+
+  @Override
+  protected Optional<ParquetValueReader<?>> timestampReader(
+      ColumnDescriptor desc, LogicalTypeAnnotation.TimeUnit unit, boolean isAdjustedToUTC) {
+    if (unit == LogicalTypeAnnotation.TimeUnit.MILLIS) {
+      return Optional.of(new ParquetValueReaders.TimestampMillisReader(desc));
+    }
+
+    return Optional.of(new ParquetValueReaders.UnboxedReader<>(desc));
+  }
+
+  @Override
+  protected Optional<ParquetValueReader<?>> uuidReader(ColumnDescriptor desc) {
+    return Optional.of(new ParquetValueReaders.UUIDReader(desc));
+  }
+
+  private static class StructLikeReader extends StructReader<StructLike, StructLike> {
+    private final GenericRecord template;
+
+    StructLikeReader(List<Type> types, List<ParquetValueReader<?>> readers, StructType struct) {
+      super(types, readers);
```

Comment:
The superclass doesn't use `types` so it doesn't need to be passed in. This isn't a big deal if you don't want to fix the superclass in this PR though.

Reply:
It touches Flink and other classes. I will handle in separate PR.